### PR TITLE
feat: Initial create

### PR DESCRIPTION
**Pull request by CptSchnitz**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,24 +1,22 @@
 {
-  "name": "service-name",
+  "name": "token-cli",
   "version": "1.0.0",
-  "description": "This is template for map colonies typescript service",
-  "main": "./src/index.ts",
+  "description": "This a cli for creating and interacting with keys and JWT tokens",
+  "main": "./index.js",
```
Comment (syncush) [inline package.json]: isnt it `./dist/index.js`?

File: tests/configurations/jest.setup.js [hunk 2]
```diff
@@ -1 +1,6 @@
-require('reflect-metadata');
+var fs = require('fs');
+var dir = '/tmp/token-cli-test-workdir';
```
Comment (syncush) [inline tests/configurations/jest.setup.js]: won't work in windows, is there a way to easily support windows %TEMP%?

File: tests/configurations/jest.teardown.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+const rimraf = require('rimraf');
+
+module.exports = function () {
+  rimraf.sync('/tmp/token-cli-test-workdir');
```
Comment (syncush) [inline tests/configurations/jest.teardown.js]: the value in jest.setup.js and jest.teardown.js should be the same, extract it to a common file.

File: tests/integration/commands/generateKey.spec.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
+// import { parseJwk } from 'jose/jwk/parse'
```
Comment (syncush) [inline tests/integration/commands/generateKey.spec.ts]: remove if not needed

File: tests/integration/commands/generateKey.spec.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      const { exitCode, stdout } = await executeCli(['generate-key', '-s', '-i', kid]);
+
+      expect(exitCode).toEqual(0);
+
+      const { privateKey, publicKey } = JSON.parse(stdout) as { publicKey: Record<string, unknown>; privateKey: Record<string, unknown> };
+      expect(privateKey).toMatchObject({ kid, alg: 'RSA256' });
+      expect(publicKey).toMatchObject({ kid, alg: 'RSA256' });
+
+      // await expect(parseJwk(privateKey)).resolves.not.toThrow();
+      // await expect(parseJwk(publicKey)).resolves.not.toThrow();
```
Comment (syncush) [inline tests/integration/commands/generateKey.spec.ts]: remove if not used

File: tests/integration/commands/generateKey.spec.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    });
+
+    it('should return a key pair in jwk format to a file', async function () {
+      const kid = 'avi';
+      const { exitCode, stdout } = await executeCli(['generate-key', '-i', kid, '-p', '/tmp/token-cli-test-workdir']);
```
Comment (syncush) [inline tests/integration/commands/generateKey.spec.ts]: throughout the code, you used hardcoded value for the files

File: tests/integration/commands/generateToken.spec.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  p: '9_CcB3B-26gyIRIxsnulTGQ4uWv7jrS0tEUJHRUL47hoAhlmzG4KQ1aK9K-qsgW7G2ovAeeCfxIgM1KKGcKXFjYzwt3fToZm-SzmcNbs6iTUCwF7ZcJvbBUPAI5hmP6uVg-XfgBapf0lfbEoS_3eyE1iLrCKl6UP_THQcCHtlos',
+  q: '0M1pdDHWU7tTSCAKm9phnPb89OuqYauBPRD7oijw08PuYvFKzY5gkDk9IISUzDYD4H7c_R2-qt9bIsVhJKE9QE9ZTJIyyd4NUhy2Z1MfkcWRb_rP9HM5keDHnI-7-SXKb6qCL-e7u50P2JKPqirlPEaRlxike-QiBNnWbY8UzW0',
+  dp: 'R9Pyv8z9U1FUxHPq5dQgPRNDw6PeyqbbOuZND5nLSftgVktZlpvCaWOm-d_ySYSTgguu9pRFH972aSfeTjwrXQny0N9KFAunxUQWtlXtaJ-mtlGFmCNN-DfZksWszI_rnY8CykP7_t5pgBJUpD81zT1t1iza2be9_BH0WqfZlqU',
+  dq: 'VT7Hgsyl9-FOiu-p5R8vIuojM8rS8LUYOloooXedepdPR83DAFDUU41Ky2_VX9ndk1W32w69vP0bNALpOg8p9otupbg6S_f_8DoQrfZduSGz0sfPBrnUEx514BE-aOVo1FYEeJJfLqVcxspvJMYGeKlG3y0MWhg1W02kIW58y7E',
+  qi: 'REh8BI-gbD_BdD2789xEoibKXjTTlSbPuf2PX9dBKsBgL7z4qlWegmLXx9uuaPHKagrbfcHvNkM2wx4telIc5nmWdvZITetxfxB2ICtbhwsOk0LIXHo26f_AvdTleicRGL0lbYc7wv71bBF4q34ro_KOOdKfophSjzElskeMiUQ',
+  alg: 'RSA256',
+  kid: 'avi',
+};
```
Comment (syncush) [inline tests/integration/commands/generateToken.spec.ts]: maybe we should store it as a test asset in a separate file, wdyt?

File: src/commands/generateKey.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      if (argv.p !== undefined && (argv.p === '' || !existsSync(argv.p))) {
+        throw new Error('the path to the files is not valid');
+      }
+      return true;
+    });
+  return yargs as yargs.Argv<GenerateKeyArguments>;
+};
+
+export const handler = async (argv: GenerateKeyArguments): Promise<void> => {
+  const { privateKey, publicKey } = await spinify(generateKeyPair, { message: 'generating key', timeout: 2000 }, argv.kid);
```
Comment (syncush) [inline src/commands/generateKey.ts]: 2 seconds is too much, reduce it, extract the value to a constant

File: src/commands/generateKey.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  const path = argv.p;
+  if (path !== undefined) {
+    try {
+      await spinify(
+        async () => {
+          await writeFile(join(path, 'publicKey.jwk'), JSON.stringify(publicKey), { encoding: 'utf-8' });
+          await writeFile(join(path, 'privateKey.jwk'), JSON.stringify(privateKey), { encoding: 'utf-8' });
```
Comment (syncush) [inline src/commands/generateKey.ts]: I think `publicKey.jwk` and `privateKey.jwk` are also a constant that should be extracted

File: src/commands/verifyToken.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+export const builder: yargs.CommandBuilder<{}, VerifyArguments> = (yargs) => {
+  yargs.option('f', { alias: ['public-key-file', 'key-file'], description: 'path to load the public key (in jwk format) from', demandOption: true });
+  yargs.option('token', { alias: 't', description: 'the token to verify', demandOption: true });
+  return yargs as unknown as yargs.Argv<VerifyArguments>;
+};
+
+export const handler = async (argv: VerifyArguments): Promise<void> => {
+  const { key: publicKey, kid } = await spinify(readAndParseJWK, { message: 'reading and parsing the public key', timeout: 1500 }, argv.f);
+  const payload = await spinify(verifyToken, { message: 'verifying token', timeout: 1000 }, publicKey, argv.token, kid);
+  console.log(JSON.stringify(payload));
```
Comment (syncush) [inline src/commands/verifyToken.ts]: it is better if we directly write to stdout and not use console.log

File: src/schemas/jwk.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+import { JTDSchemaType } from 'ajv/dist/core';
```
Comment (syncush) [inline src/schemas/jwk.ts:2]: can we import if rom `ajv` and not `ajv/dist/core`

File: tests/helpers/execute.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+export const executeCli = async (args: string[] = [], opts: { env: Env } = { env: undefined }): Promise<ExecuteReturn> => {
+  const { env = undefined } = opts;
+  const childProcess = createProcess('./src/index.ts', args, env);
+  childProcess.stdin.setDefaultEncoding('utf-8');
+  const promise = new Promise<ExecuteReturn>((resolve, reject) => {
+    // childProcess.stderr.once('data', (err: string) => {
+    //   reject(err.toString());
+    // });
```
Comment (syncush) [inline tests/helpers/execute.ts]: remove if not needed

File: tests/integration/commands/generateKey.spec.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+      expect(exitCode).toEqual(0);
+      expect(stdout).toEqual('');
+
+      const privateKey = JSON.parse((await readFile('/tmp/token-cli-test-workdir/privateKey.jwk')).toString()) as unknown;
+      const publicKey = JSON.parse((await readFile('/tmp/token-cli-test-workdir/publicKey.jwk')).toString()) as unknown;
+      expect(privateKey).toMatchObject({ kid, alg: 'RSA256' });
+      expect(publicKey).toMatchObject({ kid, alg: 'RSA256' });
+
+      // await expect(parseJwk(privateKey)).resolves.not.toThrow();
+      // await expect(parseJwk(publicKey)).resolves.not.toThrow();
```
Comment (syncush) [inline tests/integration/commands/generateKey.spec.ts]: remove if not used

File: tests/integration/commands/generateKey.spec.ts [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+      const { exitCode, stdout } = await executeCli(['generate-key', '-i', kid]);
+
+      expect(exitCode).toEqual(0);
+
+      const { privateKey, publicKey } = JSON.parse(stdout) as { publicKey: Record<string, unknown>; privateKey: Record<string, unknown> };
+      expect(privateKey).toMatchObject({ kid, alg: 'RSA256' });
+      expect(publicKey).toMatchObject({ kid, alg: 'RSA256' });
+
+      // await expect(parseJwk(privateKey)).resolves.not.toThrow();
+      // await expect(parseJwk(publicKey)).resolves.not.toThrow();
```
Comment (syncush) [inline tests/integration/commands/generateKey.spec.ts]: remove if not used

**Review by syncush — COMMENTED** [review]
1. I think the tests are missing an e2e flow

File: package.json (hunk 1 above)
Comment (CptSchnitz) [reply package.json]: y

File: tests/configurations/jest.setup.js (hunk 2 above)
Comment (CptSchnitz) [reply tests/configurations/jest.setup.js]: changes to use `os.tmpDir()`

File: tests/configurations/jest.teardown.js (hunk 3 above)
Comment (CptSchnitz) [reply tests/configurations/jest.teardown.js]: extracted

File: tests/integration/commands/generateKey.spec.ts (hunk 4 above)
Comment (CptSchnitz) [reply tests/integration/commands/generateKey.spec.ts]: fixed

File: tests/integration/commands/generateKey.spec.ts (hunk 5 above)
Comment (CptSchnitz) [reply tests/integration/commands/generateKey.spec.ts]: fixed

File: tests/integration/commands/generateKey.spec.ts (hunk 6 above)
Comment (CptSchnitz) [reply tests/integration/commands/generateKey.spec.ts]: not anymore

File: tests/integration/commands/generateToken.spec.ts (hunk 7 above)
Comment (CptSchnitz) [reply tests/integration/commands/generateToken.spec.ts]: agreed

File: src/commands/generateKey.ts (hunk 8 above)
Comment (CptSchnitz) [reply src/commands/generateKey.ts]: lowered to 1500ms

File: src/commands/generateKey.ts (hunk 9 above)
Comment (CptSchnitz) [reply src/commands/generateKey.ts]: fixed

File: src/commands/verifyToken.ts (hunk 10 above)
Comment (CptSchnitz) [reply src/commands/verifyToken.ts]: fixed

File: src/schemas/jwk.ts (hunk 11 above)
Comment (CptSchnitz) [reply src/schemas/jwk.ts:2]: It did not work, tests failed with the different import path

File: tests/helpers/execute.ts (hunk 12 above)
Comment (CptSchnitz) [reply tests/helpers/execute.ts]: removed

File: tests/integration/commands/generateKey.spec.ts (hunk 13 above)
Comment (CptSchnitz) [reply tests/integration/commands/generateKey.spec.ts]: removed

File: tests/integration/commands/generateKey.spec.ts (hunk 14 above)
Comment (CptSchnitz) [reply tests/integration/commands/generateKey.spec.ts]: removed

File: tests/integration/flow.spec.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
+import fs from 'fs';
+import path from 'path';
+import { executeCli } from '../helpers/execute';
+import { CLIENT_NAME, FILE_STORAGE_DIR, KID } from '../testConstants';
+
+const flowWorkWorkdir = path.join(FILE_STORAGE_DIR, 'flow');
+const flowPublicKeyPath = path.join(flowWorkWorkdir, 'publicKey.jwk');
+const flowPrivateKeyPath = path.join(flowWorkWorkdir, 'privateKey.jwk');
+
+jest.setTimeout(20000);
```
Comment (syncush) [inline tests/integration/flow.spec.ts]: wdyt about setting a timeout for each test instead of setting a global timeout in module level for jest?

Comment (CptSchnitz) [reply tests/integration/flow.spec.ts]: changed

**Review by syncush — APPROVED** [review]
Please add commands to the scopes of the commitlint

**Pull request merged**